### PR TITLE
k8s-friendly dep

### LINKED ISSUE #61490: RFC: dep support
So I have been working on a way to make dep (or something dep-compatible) friendlier towards the kubernetes main repository.

It differs from #57795 in that:
- it doesn't leave dep entirely as-is (instead, it tries to teach it to handle mono-repositories better)
- it doesn't introduce a runtime construct

After talking with @sttts I opened golang/dep#1764 which explores *one* of the possibilities, namely merging some of the logic into dep (certainly not as-is though).
I also have a version that's a standalone tool vendoring dep, in case upstreaming is not an option.
Realistically some of the logic *might* make it, while the rest would probably have to be handled in a k8s-specific tool.
The design details are entirely shared between the two options, so I'll leave golang/dep#1764 as the source of truth for now.

I'm opening this particular issue to
  * make sure I'm talking to both communities in a consistent way
  * collect kubernetes-specific feedback on the viability of the approach

@thockin @jessfraz @mattfarina @sttts (and whoever else might be interested), I'd love to get your thoughts !

thanks!

### PR DESCRIPTION
**Pull request by sigma**
**What this PR does / why we need it**:

This is a followup on #61490

This PR introduces a vendored `dep` [fork](https://github.com/sigma/dep/tree/kdep-tool) that's a superset of the original one, to make it friendlier to the Kubernetes use-cases.

It has the following main extra features (plus a few bugfixes that really should go upstream):
* handles the mono-repository structure (main repo + staging repositories) in a unified dependency graph
* handles "normal" repositories as well, falling back to regular dep logic (so that it could also be used as a drop-in replacement for test-infra, repo-infra, ...)
* removes more useless files (including some otherwise problematic test data)
* optionally generates a fake `Godeps.json` in the main repo and all staging repos for compatibility with existing scripts
* knows to ignore particular build tags and not pull corresponding dependencies
* preserve vendor/OWNERS

notes: 
* for the sake of maintaining a decent transition path, a lot of overrides are introduced. This makes it easier to force-keep things in sync between `godep` and `dep`. If we were to merge something like this, followup tasks would then include removing those overrides, and  replacing them with constraints where appropriate.
* an exception is the `github.com/pelletier/go-toml` package which *has* to move to version `1.1.0` to allow `dep` itself to build properly. Given that it's used only as part of the e2e tests usage of viper, it should be low-risk.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #61490

**Special notes for your reviewer**:

The whole `dep` vs `vgo` story is messy right now, and might require consideration. AFAICT `vgo` is nowhere near ready (some parts of it just don't exist at this point, even as a prototype) and `dep` development has slowed-down considerably since `vgo`'s annoucement. So while I'm still looking at some PRs to integrate things upstream, I'm reluctant just waiting the whole thing out (in particular golang/dep#1764 doesn't seem to go anywhere for now).

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by sigma** [comment]
@thockin @jessfraz @mattfarina @sttts @fejta @cblecker
@kubernetes/sig-testing-proposals

File: staging/src/k8s.io/api/Gopkg.toml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[[constraint]]
```
Comment (sttts) [inline staging/src/k8s.io/api/Gopkg.toml]: the Godeps.json files in the staging dirs have not changed? Do we still generated them?

Comment (sttts) [reply staging/src/k8s.io/api/Gopkg.toml]: We have to because the publishing-bot depends on them, and people vendoring with godep or glide depend on them.

Comment (sttts) [reply staging/src/k8s.io/api/Gopkg.toml]: > But it seems you're saying we're also going to need valid Godeps.json files no matter what.

I think until vgo is official and adopted wide enough, we will have to keep Godeps.json around, largest common denominator. It's sad, yes.

Potentially we can simulate a `godep restore` run based on `gopkg.lock` in the publisher bot, i.e. a run which populates the GOPATH.

Comment (sigma) [reply staging/src/k8s.io/api/Gopkg.toml]: I tried a few things, and settled for a feature in `dep` to generate all those files after a solution is found (at this point we have an exact knowledge of what packages are needed at what version for what `k8s.io/*` repo). I guess that would mean we could retire `hack/update-staging-godeps-dockerized.sh` ?

Incidentally it seems the currently generated `Godeps.json` can have duplicated entries, which is slightly weird.

Comment (sttts) [reply staging/src/k8s.io/api/Gopkg.toml]: > duplicated entries

You mean multiple packages per repo? That's normal for Godep. And the bot makes use of that to avoid running godep safe+restore when nothing changed. So I fear we need that as well.

Comment (sigma) [reply staging/src/k8s.io/api/Gopkg.toml]: no I mean multiple entries for the same package. See the diff for `staging/src/k8s.io/api/Godeps/Godeps.json` for example. The removed entries are the ones that were previously duplicated.

Comment (sttts) [reply staging/src/k8s.io/api/Gopkg.toml]: That's indeed weird.

Comment (sttts) [reply staging/src/k8s.io/api/Gopkg.toml]: The generated Godeps.json looks fine now. The publishing bot should work with that.

File: staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  branch = "master"
+  name = "k8s.io/apimachinery"
+
+[[constraint]]
+  branch = "master"
+  name = "k8s.io/apiserver"
+
+[[constraint]]
+  name = "k8s.io/client-go"
+  version = "6.0.0"
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml]: where does this file come from? Is it manually written?

Asking because this must be kept up-to-date. At least we need some hack/verify-* script to check that no constraint is missing

Do we really need the branch and version constraints here? Won't your dep fork figure out that the staging packages are in the same repo?

Background: on publishing we will drop this file (via https://github.com/kubernetes/publishing-bot/blob/master/configs/kubernetes-rules-configmap.yaml#L9) for now. 

I tried to keep the file in https://github.com/kubernetes/publishing-bot/pull/55 with full dependencies and hit a wall with the solver. Maybe we can do a minimal file like you have here. But it has to be generated by the bot automatically from the branch config.

Comment (sigma) [reply staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml]: Sorry, those are leftovers from previous iterations. I think at this point, those should mostly be empty and it's probably easier for all constraints to be expressed in the main Gopkg.toml anyway.
I guess I could implement some smarts that would replicate the constraints that make sense in each sub-project, but I'm not convinced that's needed for now.

I suppose I should consider generating the corresponding Gopkg.lock files though...

Comment (sigma) [reply staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml]: btw my experience is that it's easier to generate a Gopkg.toml using overrides rather than constraints when you know ahead of time what the desired solution is.
Of course, the result is not flexible, but since it'll be re-generated when needed anyway, it's probably no big deal.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml]: > rather than constraints when you know ahead of time what the desired solution is.

And overrides are ignored when they are in a library. That's the reason they are of no use in client-go. Also the lock files is ignored.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml]: To help users with dep though, we could create a Gopkg.toml.template file. They can copy that into their projects, with a long list of overrides. Wdyt? Better than nothing.

Comment (sigma) [reply staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml]: Oh sorry I misunderstood the problem at hand.
Yeah I guess providing a template might help somewhat, although it just pushes the problem one step further, as you note.

**Comment by BenTheElder** [comment]
/test pull-kubernetes-node-e2e
verify failures look to mostly be legitimate, though the typechecker may need some updates regarding staging workarounds.

File: staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml (hunk 2 above)
Comment (sigma) [reply staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml]: btw I should add that it's entirely possible that the current "solution" expressed in Godeps.json is not an actual solution and that packages would conflict with each other according to their respective constraints (which overrides casually ignore).
I don't know if that's the case currently, but a few weeks back, there was a clear inconsistency for some microsoft-related packages (whether the corresponding constraints were faulty or too broad is another question :))

My point is, in general we shouldn't assume that a godep-snapshot is dep-consistent, so generating one from the other is a risky endeavor by nature.

**Comment by BenTheElder** [comment]
node-e2e is a widespread flake, I've pinged sig-node about looking into this.
/test pull-kubernetes-node-e2e

**Comment by sigma** [comment]
@BenTheElder thanks for the update. And I agree that the verify failures are legit, I'm looking into them (none are really unexpected at this point)

File: staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml]: Godep does no consistency checks. It only checks out what you declare in Godeps.json. So if kubernetes itself (all of it) builds and works with the given revisions, so will the staging repo.

If I understand you correctly, you are referring to constraints in dependencies' Gopkg.toml. The Godep.json solution might now matchs of the dependencies in this sense. This of course can happen.

Do I understand it correctly that the generated Godeps.jsons of all staging repos agree on the revisions? I.e. you don't try to find solutions for each staging dir alone, but one global one.

Comment (sigma) [reply staging/src/k8s.io/apiextensions-apiserver/Gopkg.toml]: Right, yes all generated here `Godeps.json` are consistent, and use the versions from the unified `Gopkg.lock`

What I'm trying to express is that this `Gopkg.lock` solution is not necessarily a "real" one, in the sense that it's not the result of the resolution of proper constraints (from kubernetes itself or any of the vendored repository that happens to have a `Gopkg.toml` too). Instead, it's a forced replica (via overrides) of whatever the current `godep` solution is, at the time of the conversion.

As such, there can be no expectation (yet) that using any of the kubernetes repositories in the context of a proper dep-managed downstream project (e.g. vendoring `client-go`) should lead to dep finding a solution of said downstream project since there might not even be a dep-computable solution for the repository itself at this point. And that's also what could happen with your solver problem above  (kubernetes/publishing-bot#55)

That's a problem that should gradually disappear after migration to dep, when we replace overrides with constraints, make the dependencies versions evolve, and hopefully converge towards an override-free manifest.
In short, as long as overrides are needed, downstream is screwed and needs to override too :( (just as it is today)

**Comment by cblecker** [comment]
@sigma You can't just remove the verify scripts. The verify that the vendor is up to date, and that the staging dependency definition files are up to date with the main file. They would need to be modified to support dep, but the step of verifying vendor is sane is still required.

**Comment by sigma** [comment]
@cblecker which is why it's a WIP and I added a TODO for myself :)
Just trying to figure out if I have a complete view of what's missing

**Comment by cblecker** [comment]
*Fixing* the rest is missing, not *removing* the test. The test job is **supposed** to fail until you actually implement a proper, working test.

**Comment by cblecker** [comment]
I know that this is still a WIP, but had another look today as it appears the tests are passing.

A few thoughts to ponder on this direction:
- godep sucks, but it's "working". We have a bunch of wrapper scripts to get it to do what we need it to. I'd love to move to dep, but having to fork it and maintain a fork of a project that we don't know the future of.. I'm questionable that that's the right direction to take. If the functionality is upstreamed, then this concern goes away.. it's just the ongoing maintenance of a fork that I'm not sold on.
- From my POV, the experimental proxy that @sttts had here (https://github.com/kubernetes/kubernetes/pull/57795) might be a preferable solution because it operates beside dep, as opposed to changing dep code itself. On the other hand, the downside of that is there's a potential a breaking change in dep could break all that code too.
- If we are still generating and maintaining Godeps.json files *anyways*, and we still need wrapper scripts *anyways*, is moving to dep really that much of an advantage over godep?
- Lastly, if we do choose to continue down this path, we would need to house this fork somewhere (k8s.io/repo-infra?), and we should change the name of the executable so that it doesn't conflict with people who want to use a pure upstream dep.

**Comment by sigma** [comment]
@cblecker I probably don't have great answers to provide (I mean, any which way I look at it, the whole story looks like a terrible hack), but let me offer a few points:
- I'd definitely like some of the ideas here make it to dep upstream. And from discussions over there, it didn't seem out of the question (the implementation would look pretty different, but let's ignore this). That was before the whole vgo debacle though, so who knows what's gonna happen when... :( Note that the current fork has been designed to be as "rebasable" as possible, so I'm fairly confident I (or anybody interested) should be able to maintain it with little overhead. I agree this is not ideal.
- Note that the proxy approach doesn't cover exactly the same ground. For the most part it's not a big deal, but arguably some bugfixes currently in the fork would need to make it upstream anyway before it can be used. They've been proposed, but are currently not merged (due to overall lack of activity upstream I guess?). I'm thinking, for example, of the purging of cyclic symlinks in test data, which breaks bazel.
- to me the biggest advantage we'd gain from dep adoption is the ability to reason about the dependency graph. As of today, nothing enforces any of the constraints that our dependencies have explicitly defined. As a result, I've seen packages that are supposedly incompatible being used together in kubernetes. It might happen to work for our use-case, but that seems to be a reasonable red flag that we're kinda silently ignoring by design today. Also, I'm hoping that having a tool that reasons about versions could act as an incentive to actually use released versions of our dependencies. As of today, it seems only around 25% or our dependencies are consumed at a proper version (as opposed to a "random" revision). If anything, a hypothetical vgo-enabled future would clearly benefit from having more versions in it (an excessive number of v0.0.0-xxxx versions makes vgo's life pretty miserable).
- I definitely agree the tool would have to be hosted somewhere sensible. My first iteration was named "kdep" (and there are still leftovers from this name here and there), I'm happy to resurrect this if deemed appropriate. I'll probably delay until we agree on a canonical import path though. It's a minor adjustment anyway I think.

**Comment by sigma** [comment]
I should probably add a few more points after the fact:
- the kdep executable is usable as is, without a wrapper script. In particular, the output of `kdep status` is (or will be, once we get rid of overrides) a useful addition to our current toolbox.
- over time I've seen multiple instances of the question "what would it take to manage k8s with dep?" (or another dependency manager). Going through that exercise provides an answer (even if not completely satisfying), in the form of a list of missing features and/or things that k8s maybe shouldn't be doing. I expect that having that answer will be beneficial for the future adoption of any other version-centric dependency manager (such as vgo). So even if we have to revisit the whole thing later, I believe we'll be better off at that point.

**Comment by sigma** [comment]
I think I've taken care of all the items on my list, so tentatively removing the WIP marker

**Comment by cblecker** [comment]
/assign @sttts @thockin @cblecker
Adding this to the review plate if this is ready for showtime.

**Comment by sigma** [comment]
So I'm having a look at the community page that would need to be written for using dep, and in particular at the workflow for updating a dependency.

Right now, `dep` doesn't support passing a version (constraint) on the command line for an update, like
`dep -v -update github.com/Azure/go-autorest@=v10.7.0`
Instead, one is expected to edit Gopkg.toml and perform an update (which is less convenient, as it provides less control over which version is used exactly when range constraints are used) 

I have yet another patch that accepts that, and computes an intersection with whatever is in Gopkg.toml (which was a TODO in the code, so not absurd). That sounds good in general, but in the spirit of keeping the fork as small as possible, I'm wondering if we'd use something like that as part of the developer workflow, or if instead we'd have only exact version constraints only anyway (which would let us stay closer to the current situation and guarantee that downstream projects cannot use versions that we have not explicitly validated), which would render the feature useless. Any opinions?

**Comment by BenTheElder** [comment]
garbage collector flakes 👿
/retest

**Comment by rsc** [comment]
Please don't do this. I and the rest of the Go team would be happy to work with you to make Kubernetes work well with vgo (eventually just "go") instead of having K-specific tools on the side. If you fork dep you're just taking on more one-off technical load and maintenance instead of being able to take advantage of the rest of the ecosystem.

**Comment by sigma** [comment]
@rsc I'm personally more than happy to discuss and figure out (and even contribute to) a way forward with vgo. As mentioned in various places, I've been playing with its current incarnation in the context of k8s and experienced no shortage of valuable pain :).
That said, I don't think it's reasonable to (ask to) reject that particular transition before it is clear when vgo will be 1. real, and 2. a good fit for k8s. The godep-induced pain is real, and needs urgent fixing IMO. Waiting only makes the situation worse, and any transition more difficult.
Also, my view expressed above is that this particular transition would actually help a future move to vgo, by virtue of simply starting to care about semver at all. Why do you feel that's not the case?

**Comment by rsc** [comment]
@sigma I'll set up a time to talk, as we discussed. I just want to reply briefly to the point you raised here, to avoid the appearance of ignoring you.

- It seems to me that writing and maintaining a custom tool is a big deal. Dep is a very complex tool, and I would very much caution against forking and taking on maintenance of a fork.
- The vgo prototype has improved dramatically in the past month. I think it's very close to being able to support Kubernetes. If you want to put time into replacing the use of godep, I think it would make more sense to jump directly to vgo than via a custom dep fork.
- I understand that use of godep has been less than ideal for Kubernetes, but I also know that it's worked well enough for a few years. That would seem to imply to me that it can serve for the couple more months needed to get over to vgo.
- Dep has enough problems with its model that I don't even think I'd suggest moving to unmodified dep as a stepping stone at this point.

**Comment by thockin** [comment]
@sigma I am not aware of acute pain wrt godep today.  The scripts we have wrapping it are slow, and occasionally generate weird deltas (e.g. one more or less character in the hashes) but they work reliably and they are not run THAT often.

@rsc I am glad to hear you even considering k8s' bizarre needs.  Is there any documentation or anything to describe the principles/patterns vgo will adopt to handle the mess we have made (e.g. staging/... as a pseudo-vendor'ed thing)?

**Comment by sigma** [comment]
@thockin for the staging/ sub-repositories, my current understanding is that the strategy would look like:
- each sub-repo is a separate vgo module
- the main repo and each sub-repo would refer to other sub-repositories using the replace mechanism of vgo, to point to the local filesystem (that part fits pretty nicely in the vgo model)
- we would have to maintain the consistency of versions across main and sub-repositories by having vgo resolve dependencies at the top-level, and then use some script to bump each dependency of each sub-repository to the same version as the one at top-level (that part doesn't fit that well)

**Comment by sigma** [comment]
@thockin @sttts @cblecker so should we close this PR and move towards vgo integration?
There's quite a lot of work that needs to happen there before it can really be used, so keeping those parallel tracks seems wasteful.

**Comment by sttts** [comment]
> so should we close this PR and move towards vgo integration?

I think vgo is the better approach long-term. @rsc's prototype looks pretty convincing. So, I vote for going the vgo route and accepting the godep pain for a bit longer.

**Comment by cblecker** [comment]
+1 to @sttts

**Comment by mattfarina** [comment]
We should head down the vgo/go modules route. Why switch twice when we can switch once? Heading that way will also enable us to provide feedback to the vgo team to make sure our use cases are being met.

**Comment by sigma** [comment]
seems like we have general agreement, and I'm not pursuing this any further anyway.
If for some reason we need to reconsider down the road, I'll reopen.

/close
